Scoping, multi file resolution bugfix

goto40 opened the pull request:
I realized some issues with multi file reference resolution.

I like to discuss the use of one separate parser per model file (independent from the fact if the models are from the same meta model).

Changes and rationale: see CHANGELOG.md.

Commit:
Fixed the "is_main_model"-flag when loading dependent model files (in
case of metamodel changes).
Fixed multi-metamodel support for Plain Name Lookup.
Fixed textx_isinstance (not 100% sure about completeness).

Commit:
Fixed the "is_main_model"-flag when loading dependent model files (in
case of metamodel changes).
Fixed multi-metamodel support for Plain Name Lookup.
Fixed textx_isinstance (not 100% sure about completeness).

Commit:
allow to use one single global repo (for different meta models)

Commit:
Changed Exception to include filename in errors.
Changed Exception __str__ function to generate gcc-like error messages.
This can be adapted later easily with custom exception-to-string functions.
Adapted some tests.

Commit:
Added more complex test. Failing for importURI. Problem: resolution order.

Commit:
Added more complex test. Prepared inhertance test. Resolution must be decoupled from model loading.
We need more than one parser in order to use it successfully for correct lin/col deduction.

Commit:
Added more complex test. Failing for concatenated inheritance. We cannot distinguish from extends==None and extends==Postponed (???) TBC

Commit:
Now we use one parser per model. With this we get correct behavior when mixing metamodels and also when distributing a model across files (col/line).

Commit:
Added example with list of references

Commit:
removed printfs

Commit:
fixes

Commit:
code style

Commit:
docu

Commit:
docu typos

Commit:
* Added tool function (get_location) to ease raising errors with correct line/col/filename info.
 * Added _tx_parser to the model and added docu (model.md) accordingly
 * model.py: added the parser to the model
 * removed the parser from the scope API --> this forces to use the parser from the model (e.g. via textx.scoping.tools.get_parser)
   and, thus, to always use the correct parser.
 * enhanced a unittest to check that a different parser ist used for each file (and fixed an error: missing clone()-call in metamodel.py)
 * added a test to demonstrate the use of get_location and to check an error raised from an included model file (with correct line/col/filename).

igordejanovic commented:
@goto40 Hi Pierre. Just to let you know that I don't have much time at my disposal at the moment but I'll review this PR as soon as I can. Thanks for the contributions and sorry for my slow responses.

goto40 commented:
Hi Igor,

No problem. Take your time. No hurry! Let's wait until there is time for
proper reviews. I prefer quality and common understanding to quick shots.

Best, Pierre

Am 27.03.2018 22:14 schrieb "Igor Dejanović" <notifications@github.com>:

> @goto40 <https://github.com/goto40> Hi Pierre. Just to let you know that
> I don't have much time at my disposal at the moment but I'll review this PR
> as soon as I can. Thanks for the contributions and sorry for my slow
> responses.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/igordejanovic/textX/pull/60#issuecomment-376659753>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQGtWjUupNAn7G_G792eeTN5NGH6O0EOks5tip2vgaJpZM4S3RMU>
> .
>

Commit:
Added diamond include test (one file included via two paths).

Commit:
docu

Commit:
enhanced export_model to support distributed models. This - however -
requries a meta model with appropriate scope provider. To be discussed:
the "textx visualize" command for models with custom scope providers
is not working...

The scope providers belong to the meta model, and thus are required to
successfully parse a model.

goto40 commented:
I also added the ability to render a model (export.model_export) which is distributed across different files. An example/test is also provided (test_scoping/test_model_export). See last commit comment...

Commit:
docu for the tests.

Commit:
Added new attribute _tx_loaded_models for objects used to load other model files.
This is mandatory for languages, where the import statement is used to place
the imported elements at a particular location (like C-#includes)...

igordejanovic commented on tests/functional/test_scoping/test_children.py:
```diff
@@ -32,7 +32,10 @@
     '''
```
I've noticed now that `RefClass` rule is not used anywhere. There is another usage in file `test_fully_qualified_name.py`

goto40 replied:
removed dead code in grammar (2x)

igordejanovic commented on tests/functional/test_scoping/test_exception_from_included_model.py:
```diff
@@ -0,0 +1,65 @@
+from __future__ import unicode_literals
+from textx import metamodel_from_file, get_children_of_type
```
There are several PEP-8 issues in this file. For example, `get_children_of_type` is not used.

igordejanovic replied:
There are some PEP-8 issues in other files too. I can fix those after the merge.

goto40 replied:
I think I fixed many of these...

igordejanovic commented on tests/functional/test_scoping/metamodel_provider3/Base.tx:
```diff
@@ -0,0 +1,11 @@
+Cls: "class" name=ID (
+( "extends" extends+=[Cls][',']+)?
```
`+` at the end of `extends+=[Cls][',']+` is superfluous as it is implied by `+=`

igordejanovic replied:
Rule `Cls` might be better indented for readability.

goto40 replied:
Ok. Fixed.

igordejanovic reviewed (changes requested):
Excellent job Pierre!
There only few small fixes. Please see inline comments.

igordejanovic commented:
@goto40 BTW, I think that this multi-(meta)model support deserves a mention in `Feature highlights` in `index.md` :)

igordejanovic commented:
@goto40 and also, multi-(meta)-model support is such an important addition to textX that it would be awesome to have a dedicated section in the docs with examples (probably some examples used in tests) and explanations.

goto40 commented:
I will do something we can discuss... Best, Pierre

Igor Dejanović <notifications@github.com> schrieb am Mo., 16. Apr. 2018,
12:11:

> @goto40 <https://github.com/goto40> and also, multi-(meta)-model support
> is such an important addition to textX that it would be awesome to have a
> dedicated section in the docs with examples (probably some examples used in
> tests) and explanations.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/igordejanovic/textX/pull/60#issuecomment-381549704>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQGtWlc-YyzlynlAJkyNDi090CexX2VYks5tpG5XgaJpZM4S3RMU>
> .
>

Commit:
removed dead code in grammar

Commit:
PEP8 corrections

Commit:
PEP8 corrections, optimize imports

Commit:
Redundant "+" rmeoved. Reformatted.

Commit:
Docu updated (scoping)

Commit:
New test illustrating how to reference into a non-textX model.

goto40 commented:
I worked in the in-line comments (done; maybe check for the PEP-stuff).

I added a new test (I understood something while reading your comments): test_reference_to_buildin_attribute.py

I did not yet work on the docu part (in progress).

goto40 commented:
I have one big issue/idea (for the next pull request) which influences in parts the multi meta model feature... I will - for the moment - ***ignore these ideas and write some documentation for the current state***, since I have many open points for the concrete design/solution of my wish...

# How to extend a Meta model defined in another (installed) python package?

***Use case***: a meta model installed via pip should be extended. The installed meta model could be part of a language (e.g. some script language). The goal is to define a new meta model extending the installed meta model. In this example this new meta model could be used be to define a test specification referencing functions defined in the installed/extended/used language. 

What should be inherited from the metal model to be extended?
  - the grammar: ok  
  - object processors (e.g. validators): must be re-added manually. 
  - scope providers (same as for object processors)

Maybe it is of interest to include (parts of) a grammar for referencing purposes only. This means to inhibit creating an instance of that grammar rule in the extended meta model and, thus, only allowing to create references for that inherited rule. The latter case then makes it unnecessary to inherit most of the scope providers stuff. 


## Grammar extension

Unclear how to find the grammar file in case of grammars to be imported from other python packages (e.g. import grammarname will not work out of the box). Maybe we should somehow add a get grammar method to the Meta model? However I do not see a nice way to use it? Or maybe we should add the possibility to modify the search path for the grammars? But in the latter case, it should also work with grammars hidden in a package archive (egg or whatever). 


## Object processors

No direct support for multiple processors for one grammar rule. Maybe we should allow lists of object processors in addition to object processors?


## Scope providers

When redefining rules (creating a rule with the same name as an inherited one), we could allow in the context of the scope provider mapping to allow to use the full rule names (Grammarname.rulename) to identify both rules separately (both: the rule in the grammar to be extended and the redefined rule). However, I think this case is quite exotic and can avoided easily.

goto40 commented:
If the last comment is confusing, ignore it for the moment, please (I just wanted to write down this ideas). Maybe it takes longer to find a proper solution... Or is it no problem at all.

I will just write some docu for the current state...

igordejanovic commented:
@goto40 I was thinking about similar ideas a lot in the past. How to deploy, find, inherit, extend existing languages in the system.

For discovering existing languages in the system [pkg_resources](http://peak.telecommunity.com/DevCenter/PkgResources) can be used. See [this little tool](https://github.com/textX-tools/textx-tools), that discovers existing languages and can create a scaffolding project for the given language. For example, [this language](https://github.com/igordejanovic/textx-lang-er) registers itself in the [setup.py here](https://github.com/igordejanovic/textx-lang-er/blob/master/setup.py#L58) and is discoverable by [textX tools](https://github.com/textX-tools/textx-tools). The tool finds all languages using [textX language API](https://github.com/igordejanovic/textX/blob/master/textx/langapi.py#L12). We could make extension points for whatever we would like to be discoverable, like meta-models, grammars etc.

Maybe it would be the best to make a new issue to discuss.

igordejanovic commented:
Looks good. Merging.
